atari-DQN/dqn_agent.py

```diff
@@ -0,0 +1,315 @@
+                                                     ).to(self.device)
+        self.target_net = network.nature_dqn_network(self.num_actions,
+                                                     self.history_size,
+                                                     self.observation_shape[-2:]
+                                                     ).to(self.device)
+        self.target_net.load_state_dict(self.policy_net.state_dict())
+        self.target_net.eval()  # set target net to evaluation mode
+
+    def _init_memory(self) -> None:
+        """Initialize the memory buffer"""
+        self.memory = replay_memory.CircularReplayBuffer(
+            buffer_cap=self.memory_buffer_capacity,
+            history=self.history_size,
+            obs_shape=self.observation_shape,
+            device=self.device)
+
+    def begin_episode(self) -> None:
+        """
+        Prepare the agent at the beginning of a new episode
+        """
+        # Initialize (zero-padded) history
+        for _ in range(self.history_size):
+            zero_pad_mat = torch.zeros(self.observation_shape,
+                                       dtype=self.observation_dtype,
+                                       device='cpu')
+            self.history_queue.append(zero_pad_mat)
+
+        # clear per-episode counters
+        self.episode_total_policy_loss = 0.0
+        self.episode_total_optim_steps = 0
+
+        # NOTE: first observation is not stored in memory buffer
+        # TODO: maybe add training step?
+
+    def step(self, observation: np.ndarray) -> int:
+        """
+        The agent takes one step
+
+        :param observation:  initial observation from the environment, should
+                have shape self.observation_shape
+        :return: int denoting action to take at this step
+        """
+        # Cast observation to tensor and add to history
+        observation = torch.tensor(observation, dtype=self.observation_dtype,
+                                   device='cpu')
+        self.history_queue.append(observation)
+
+        self._train_step()
+        # Select action and return int
+        return self._select_action()
+
+    def _select_action(self) -> int:
+        """
+        Select action according to the epsilon greedy policy
+        :return: int denoting action to take
+        """
+
+        # Compute epsilon
+        epsilon = self.epsilon_fn(self.epsilon_decay_period,
+                                  self.training_steps,
+                                  self.min_replay_history,
+                                  self.epsilon_final)
+
+        if random.random() <= epsilon:
+            # random action with probability epsilon
+            return random.randrange(self.num_actions)
+        else:
+            # Construct state
+            state = self._history_queue_to_state()
+            # greedy action
+            action_tensor = self.policy_net(state).max(1)[1].view(1, 1)
```

Comment:
Shouldn't there be a `with torch.no_grad():` here before using the `policy_net` to evaluate Q value?